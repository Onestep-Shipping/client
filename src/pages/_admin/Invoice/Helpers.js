import PropTypes from 'prop-types';
import React from 'react';
import Utils from '../../../utils/Helpers.js';

const FinanceRow = props => {
  const { label, value } = props;

  return (
    <div className="invoice-row-small">
      <text className="info-label-special">{label}: </text>
      <text>${Utils.comma(value)}</text>
    </div>
  ) 
}

const NumberInput = props => {
  const { ind, onChange, fees } = props;

  return (
    <div className="usd-input-container">
      <input 
        type="number" name="price" className="usd-input" 
        onChange={e => onChange(e, ind)} required />
      <input 
        type="text" name="total" 
        className="usd-input" 
        disabled={true} 
        value={parseFloat(fees[ind]) ? Utils.comma(fees[ind]) : 0}
      />
    </div>
  );
}

export { FinanceRow, NumberInput };

FinanceRow.propTypes = {
  label: PropTypes.string,
  value: PropTypes.number
};

NumberInput.propTypes = {
  ind: PropTypes.number,
  onChange: PropTypes.func,
  fees: PropTypes.array,
};