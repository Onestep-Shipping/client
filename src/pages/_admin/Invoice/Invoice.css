.invoice-form-container {
  margin-top: 30px;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 30px;
}

.invoice-row {
  width: 80%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 30px;
}

.invoice-row-right {
  width: 80%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 30px;
}

.usd-input, .total-input {
  font-size: 16px;
  padding: 8px 15px;
  border-radius: 4px;
  border: none;
  outline: none;
  border: solid #474747 1px;
  background-color: white;
  box-shadow: none;
  color: black;
}

.usd-input {
  width: 35%;
}

.total-input {
  width: 20%;
}

.usd-input-container {
  width: 50%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
}