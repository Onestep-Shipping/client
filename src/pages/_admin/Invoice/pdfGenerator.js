import Utils from '../../../utils/Helpers.js';
import jsPDF from 'jspdf';
import moment from 'moment';

const doc = new jsPDF("p", "pt", 'letter');

const centeredText = (text, y) => {
    let textWidth = doc.getStringUnitWidth(text) * doc.internal.getFontSize() / doc.internal.scaleFactor;
    let textOffset = (doc.internal.pageSize.width - textWidth) / 2;
    doc.text(textOffset, y, text);
}

const preview = info => {
  doc.setProperties({
    title: 'Invoice #' + info.bookingNo,
    subject: 'Info about PDF',
    author: 'Scarlet Nguyen',
    keywords: 'generated, javascript, web 2.0, ajax',
    creator: 'OneStep Shipping'
  });

  doc.setFont('helvatica');
  doc.setFontSize(16);
  doc.setFontType('bolditalic');
  doc.text(430, 60, "OneStep Shipping");

  doc.setFont('courier');
  doc.setFontSize(20);
  doc.setFontType('bold');
  centeredText('Invoice', 90);

  doc.setFontSize(12);
  doc.setFontType('normal');
  doc.text(50, 130, "TO: " + info.company.name);
  doc.text(50, 148, "    " + info.company.address.street);
  doc.text(50, 163, "    " + info.company.address.city);
  doc.text(50, 178, "    " + info.company.address.country);

  doc.setFontSize(10);
  doc.text(310, 115, "Invoice No.      :                3242356");
  doc.text(310, 130, "Our Reference No.:                1234567");
  doc.text(310, 145, "Contact          :              Rose Phan");
  doc.text(310, 160, "Tel              :      +1 (604) 500-4934");
  doc.text(310, 175, "Email            :   rosephan99@gmail.com");
  doc.text(310, 190, "Issue Date       :             " + moment().format("L"));

  doc.line(50, 200, 560, 200);

  doc.setFontSize(12);
  doc.setFontType('bold');
  doc.text(80, 230, "Booking No.: " + info.bookingNo);

  doc.setFontSize(11);
  doc.setFontType('normal');

  doc.setFontSize(11);
  doc.text(80, 260, "From: " + info.schedule.route.startLocation);
  doc.text(350, 260, "To: " + info.schedule.route.endLocation);

  doc.text(80, 280, "ETD: " + Utils.formatISOString(info.schedule.startDate));
  doc.text(350, 280, "ETA: " + Utils.formatISOString(info.schedule.endDate));

  doc.text(80, 320, "Commodity: " + info.booking.commodity);
  doc.text(350, 320, "Order Number: " + info.orderNo);

  doc.setFontType('bold');
  doc.text(80, 360, "Description");
  doc.text(200, 360, "Q'ty");
  doc.text(250, 360, "Container");
  doc.text(350, 360, "Price (USD)");
  doc.text(450, 360, "Total (USD)");

  doc.setFontType('normal');

  info.priceList.map((row, i) => {
    const height = 377 + (i * 17);
    if (i < info.priceList.length - 2) {
      doc.text(80, height, "Ocean Freight");
      doc.text(200, height, "" + info.booking.containers[i].quantity);
      doc.text(250, height, info.booking.containers[i].containerType);
    } else if (i === info.priceList.length - 2) {
      doc.text(80, height, "Document");
      doc.text(200, height, "1");
    } else {
      doc.text(80, height, "Administration");
      doc.text(200, height, "1");
      doc.line(80, height + 10, 560, height + 10);
      doc.text(350, height + 24, "Sub Total: ");
      doc.text(450, height + 24, "$" + info.subTotal);
    }
    doc.text(350, height, "$" + row.price);
    doc.text(450, height, "$" + row.total);
  })

  doc.setFontSize(12);
  doc.setFontType('bold');
  doc.text(80, 530, "Bank Information - RBC Bank");

  doc.setFontSize(11);
  doc.setFontType('normal');
  doc.text(80, 550, "Branch No.       : 12345");
  doc.text(80, 567, "Institution No.  : 003");
  doc.text(80, 584, "Account No. (USD): 1234567");
  doc.text(80, 601, "Account No.      : 2345678");

  doc.line(50, 625, 560, 625);
  doc.setFontSize(11);
  doc.setFontType('normal');
  doc.text(50, 650, 'Please make Transfer to OneStep Shipping as soon as you receive this invoice.');

  doc.text(50, 690, 'Thank you & Best regards,');
  doc.text(50, 710, 'OneStep Shipping.');

  window.open(doc.output('bloburl'), '_blank');
}

const uploadToServer = () => {
  return doc.output('blob');
}

export default { preview, uploadToServer };