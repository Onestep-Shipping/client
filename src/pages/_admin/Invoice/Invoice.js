import './Invoice.css';

import { FinanceRow, NumberInput } from './Helpers';
import React, { useState } from 'react';

import BookingDisplay from '../../../components/BookingDisplay/BookingDisplay.js';
import { CONTAINER_TYPES } from '../../../constants/ServiceFormConstants';
import DATA from '../../../data/ScheduleDetailsData.js';
import FileUploadService from '../../../services/FileUploadService.js';
import Header from '../../../components/Header/Header.js';
import INVOICE from '../../../data/InvoiceData.js';
import { InfoRow } from '../Helpers.js';
import PdfGenerator from './pdfGenerator.js';
import QUOTE_DATA from '../../../data/QuoteUpdateData';
import { QuoteRow } from '../QuoteUpdate/Helpers';
import UserList from '../../../components/UserList/UserList.js';
import Utils from '../../../utils/Helpers.js';

const initCost = ind => {
  let finalCost = 0;
  INVOICE[ind].booking.container.map(row => 
    CONTAINER_TYPES.map((type, type_ind) => function() { 
      if (type === row.containerType) {
        finalCost += (QUOTE_DATA[ind].buying.oceanFreight[type_ind] * row.quantity);
      }
    })
  )
  finalCost += QUOTE_DATA[ind].buying.docFee + QUOTE_DATA[ind].buying.adminFee;
  
  return finalCost;
}

const Invoice = () => {
  const [currentIndex, setCurrentIndex] = useState(0);
  const [fees, setFees] = useState(
    Array(INVOICE[currentIndex].booking.container.length + 2).fill(0)
  );
  const cost = initCost(currentIndex);
  const revenue = fees.reduce((a, b) => a + b, 0);
  const profit = revenue - cost;


  const handleIndexChange = newInd => {
    setCurrentIndex(newInd);
    setFees(Array(INVOICE[newInd].booking.container.length + 2).fill(0));
    document.getElementsByName("price").forEach(node => node.value = "");
  }

  const handlePreview = e => {
    e.preventDefault();
    PdfGenerator.preview(createInfoObject(e));
  }

  const handleUpload = e => {
    e.preventDefault();
    const blob = PdfGenerator.uploadToServer();
    const pdf = new File(
      [blob], 
      'Invoice #' + INVOICE[currentIndex].id + '.pdf', 
      { type: 'application/pdf' } 
    )
    const formData = new FormData();
    formData.append("file", pdf);

    FileUploadService.uploadFile(formData)
      .then(res => {
        alert(res.data.fileLocation);
      })
      .catch(e => {
        console.log(e.response);
      });
  }

  const onFeeChange = (e, i) => {
    let newFees = [...fees]; 
    let newVal = parseInt(e.target.value);
    if (i < fees.length - 2) {
      newVal *= INVOICE[currentIndex].booking.container[i].quantity;
    }
    newFees[i] = newVal;
    setFees(newFees);
  }

  const createInfoObject = e => {
    const { price, total } = e.target;
    const priceList = toArray(price);
    const totalList = toArray(total);
    const combine = [];
    priceList.map((p, ind) => combine.push({ price: p, total: totalList[ind] }))
    
    return { 
      company: INVOICE[currentIndex].company,
      schedule: DATA[currentIndex],
      bookingNo: INVOICE[currentIndex].id,
      booking: INVOICE[currentIndex].booking,
      orderNo: INVOICE[currentIndex].bol.orderNo,
      priceList: combine, 
      subTotal: Utils.comma(revenue) 
    }
  }

  const toArray = nodelist => {
    let list = [];
    nodelist.forEach(node => list.push(node.value));
    return list;
  }

  return (
    <div className="homepage-container">
      <Header />
      <div className="bol-instruction-container">
        <UserList 
          setInd={handleIndexChange}
          opt={INVOICE} type="bol" />
        <div className="bol-instruction-detail"> 
          <div className="booking-id-container">
            <h1>Invoice</h1>
          </div>
          <div className="customer-info-container">
            <span>Company: {INVOICE[currentIndex].company.name}</span>
            <span>Email: {INVOICE[currentIndex].email}</span>
          </div>

          <div className="form-container">
            <h2>Schedule</h2>
            <BookingDisplay id={'' + currentIndex} fields={8}/>
          </div>

          <div className="form-container">
            <h2>Information</h2>
            <div className="booking-details-container">
              <InfoRow label="Booking Number" value={INVOICE[currentIndex].id} />
              <InfoRow label="Commodity" value={INVOICE[currentIndex].booking.commodity} />
              <InfoRow label="Order/PO Number" value={INVOICE[currentIndex].bol.orderNo} />
            </div>
          </div>

          {!INVOICE[currentIndex].isCompleted &&
          <form className="invoice-form-container" onSubmit={handlePreview}>
            <h2>Invoice</h2>
            <div className="invoice-row">
                <span className="info-label-special">Description</span>
                <span className="info-label-special">Quantity</span>
                <span className="info-label-special">Container</span>
                <div className="usd-input-container">
                  <span className="info-label-special">Price (USD)</span>
                  <span className="info-label-special">Total (USD)</span>
                </div>
            </div>
            {INVOICE[currentIndex].booking.container.map((row, ind) => 
                <div className="invoice-row" key={ind}>
                  <span>Ocean Freight</span>
                  <span>{row.quantity}</span>
                  <span>{row.containerType}</span>
                  <NumberInput ind={ind} onChange={onFeeChange} fees={fees}/>
                </div>
            )}
            <div className="invoice-row">
              <span>Document Fee</span>
              <NumberInput ind={fees.length - 2} onChange={onFeeChange} fees={fees}/>
            </div>
            <div className="invoice-row">
              <span>Administration Fee</span>
              <NumberInput ind={fees.length - 1} onChange={onFeeChange} fees={fees}/>
            </div>
            <FinanceRow label="Revenue" value={revenue} />
            <FinanceRow label="Cost" value={cost} />
            <FinanceRow label="Profit" value={profit} />
            <div className="invoice-row">
              <QuoteRow header="Buying" obj={QUOTE_DATA[currentIndex].buying} />
              <QuoteRow header="Selling" obj={QUOTE_DATA[currentIndex].selling} />
            </div>
            <div className="bol-button-form">
              <input id="left-button" type="submit" className="result-button" value="Generate PDF" />
              <button  className="result-button" onClick={handleUpload}>
                Send to Customer
              </button>
            </div>
          </form>}
        </div>
      </div>
    </div>
  );
};

export default Invoice;
