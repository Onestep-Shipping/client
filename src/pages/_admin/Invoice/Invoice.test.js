import { ApolloProvider } from '@apollo/react-hooks';
import Invoice from './Invoice';
import React from 'react';
import { shallow } from 'enzyme';

let wrapper = shallow(
  <ApolloProvider>
    <Invoice />
  </ApolloProvider>
);

describe('Invoice', () => {
  it('should render the Invoice Component correctly', () => {   
    expect(Invoice).toMatchSnapshot();
  });
});