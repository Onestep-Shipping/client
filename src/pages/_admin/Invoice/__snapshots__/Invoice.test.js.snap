// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice should render the Invoice Component correctly 1`] = `[Function]`;
